### PR TITLE
Home Network Settings

### PR DESCRIPTION
**Pull request by JBassett**
Fixes: https://github.com/home-assistant/home-assistant-android/issues/30

Things left to do:
- [x] Better handling about what error is sending us to settings, currently to sensitive
- [x] Validate that SSID checking is working (I don't think it is yet...)

### CODE REVIEW

**Comment by andriej** [comment]
Nice to see this PR - one question from an observer: will app be able to ignore 'home'/'remote'?
I have my router setup so it resolves external domain to my local IP when connected with home-wifi.
I'd like to see app not break things and be able to set just 'use one URL always'.

Just my 5 cents.

**Comment by JBassett** [comment]
I too have a router that can handle it.  The ability to use a single URL will always work.

**Comment by andriej** [comment]
Just wanted to make sure.

**Comment by quthla** [comment]
It probably needs [this](https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback) to reload the webview if needed

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt [hunk 1]
```diff
@@ -25,8 +25,8 @@ class WebViewPresenterImpl @Inject constructor(
 
     override fun onViewReady() {
         mainScope.launch {
-            val url = integrationUseCase.getUiUrl(false)
-                ?: throw IllegalStateException("Unable to display the webview if we do not have url")
+            val ssid = view.getCurrentSsid()
+            val url = urlUseCase.getUrl("\"$ssid\"" == urlUseCase.getHomeWifiSsid())
```
Comment (balloob) [inline app/src/main/java/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt]: Why the extra quotes ?

Comment (quthla) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt]: https://developer.android.com/reference/android/net/wifi/WifiInfo.html#getSSID()

File: data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+
+class UrlRepositoryImpl @Inject constructor(
+    @Named("url") private val localStorage: LocalStorage
+) : UrlRepository {
+
+    companion object {
+        private const val PREF_CLOUD_URL = "cloud_url"
```
Comment (balloob) [inline data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt]: What is `CLOUD_URL` ? There is a Remote UI cloud url and there is a cloudhook (cloud webhook) url.

Comment (balloob) [reply data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt]: I see now that it's for cloudhook, let's rename it

File: data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        localStorage.putString(PREF_WEBHOOK_ID, webhookId)
+    }
+
+    override suspend fun getUrl(isInternal: Boolean): URL? {
+        val internal = localStorage.getString(PREF_LOCAL_URL)?.toHttpUrlOrNull()?.toUrl()
+        val external = localStorage.getString(PREF_REMOTE_URL)?.toHttpUrlOrNull()?.toUrl()
+        return if (isInternal) {
+            internal ?: external
+        } else {
+            external ?: internal
```
Comment (balloob) [inline data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt]: If we're external, would it really help in ever returning the internal url ?

Comment (JBassett) [reply data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt]: It really depends on peoples setups, if someone manually sets up the url during onboarding that gets persisted as the internal address.  However, without the wifi ssid set they are technically external.  So in that case they need the internal address returned.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt (hunk 1 above)
Comment (JBassett) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt]: But I have my logic reversed...

**Comment by JBassett** [comment]
Settings screenshot:
Screenshot_20191230-082941

File: app/src/main/java/io/homeassistant/companion/android/onboarding/authentication/AuthenticationPresenterImpl.kt [hunk 4]
```diff
@@ -24,7 +24,7 @@ class AuthenticationPresenterImpl @Inject constructor(
 
     override fun onViewReady() {
         mainScope.launch {
-            view.loadUrl(authenticationUseCase.buildAuthenticationUrl(AUTH_CALLBACK).toString())
+            view.loadUrl(authenticationUseCase.buildAuthenticationUrl(false, AUTH_CALLBACK).toString())
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/onboarding/authentication/AuthenticationPresenterImpl.kt]: - [ ] //TODO: Fix me to use correct url based on WiFi Status.

**Comment by balloob** [comment]
For the screenshot, let's call external url just Home Assistant url. Preferably there would be a new section with a toggle to "Enable different url when at home".

**Comment by balloob** [comment]
Or instead of toggle maybe when no ssid typed in we disable local url field.

**Comment by JBassett** [comment]
Invalid url:
Screenshot_20191231-141441

Disabled preference when no WiFi set:
Screenshot_20191231-131820

**Comment by balloob** [comment]
🎉 

Let's test this for a couple of days and do a new release soon.

**Comment by dgtal1** [comment]
Can you allow multiple SSIDs? I've got 2.4 and 5GHz network at home, both with different SSID. Would be great to see such feature. It may be comma delimited or a list box. Great work guys!

**Comment by JBassett** [comment]
@dgtal1 there is already a request for that. https://github.com/home-assistant/home-assistant-android/issues/250

**Pull request merged**